### PR TITLE
Replace OS-based inbox with "programmatic" inbox

### PR DESCRIPTION
**Pull request by dtitov**
Replacing Unix SSHD with Apache Mina SSHD (https://mina.apache.org/sshd-project/):
- no need for PAM code (C);
- no need for FUSE or cron as file events are handled out of the box;
- possibility to integrate `verify` feature to this microservice;
- possibility to make on-the-fly files streaming to S3 without intermediate storage.

Sources to review: https://github.com/NBISweden/LocalEGA-inbox

At the moment it doesn't include ingestion functionality (doesn't put the file directly to S3), but rather just replaces existing inbox (the "dummy" one). But later if we implement direct streaming to S3, the ingestion functionality will be absorbed by that.

### CODE REVIEW

File: deployments/docker/bootstrap/lega.sh [hunk 1]
```diff
@@ -227,29 +227,16 @@ services:
...
+      - "${DOCKER_PORT_inbox}:2222"
     container_name: inbox
-    image: nbisweden/ega-inbox
-    # privileged, cap_add and devices cannot be used by docker Swarm
-    privileged: true
-    cap_add:
-      - ALL
-    devices:
-      - /dev/fuse
+    image: nbisweden/ega-mina-inbox
```
Comment (blankdots) [inline deployments/docker/bootstrap/lega.sh:229]: Would it be possible to place the `Dockerfile` in https://github.com/NBISweden/LocalEGA-inbox or even in the `deployments/docker/images` in this repo. It helps with knowing what is in inside that image.

Comment (dtitov) [reply deployments/docker/bootstrap/lega.sh:229]: @blankdots, there’s no Dokerfile: image is autogenerated by Maven plugin (see the `pom.xml`). Same approach as in Data API.

Comment (blankdots) [reply deployments/docker/bootstrap/lega.sh:229]: Noted, will take a look at the code for inbox.
Also I would not replace the whole inbox maybe, add some sort of flag in the bootstrap to build one or the other, I still view this as an alternative not a full replacement (a subject for a discussion on slack).

**Comment by blankdots** [comment]
This has my blessing :church:, functionality seems to be there.
Docs to be updated in a different PR.
@juhtornr please also review LocalEGA-inbox code then: :package: :rocket:

**Review by juhtornr — APPROVED** [review]
Works as advertised: replaces the sftpd with Apache Mina and which sents a message to RabbitMQ when file is uploaded. Should also point out that the logic is new repository: https://github.com/NBISweden/LocalEGA-inbox

**Comment by silverdaz** [comment]
It sends a message to the broker on close, but it is a slightly wrong logic:
The "on close" event should not be solely caught. It must be captured along with the create/append event. Simply reading the file will trigger an open/close sequence that will reprocess the file.
See [here](https://github.com/NBISweden/LocalEGA-inbox/blob/master/src/main/java/se/nbis/lega/inbox/sftp/InboxSftpEventListener.java#L43-L53)

**Comment by dtitov** [comment]
I'll check if reading the file triggers it, thanks.

**Comment by dtitov** [comment]
You were right. I'll make a PR to fix it soon.

**Comment by silverdaz** [comment]
We have faced these issues before.
This is redoing work that was done before.
What if I had not looked at it (out of curiosity this morning) ?

**Comment by dtitov** [comment]
1. This is a bug (non-critical bug though, in my mind). [We had bugs before](https://github.com/NBISweden/LocalEGA/issues?utf8=%E2%9C%93&q=is%3Aissue++label%3Abug++) and we will have them in the future - it's absolutely natural.
2. I'm not an oracle, but I would assume that if you haven't had looked at it this morning, we would still find this bug some time later. For instance, I had in mind to make some tests of what happens if a file is uploaded and then removed from the inbox (and some related scenarios e.g. file copying/moving, etc.), maybe even create new tests in the suite for these scenarios. 
3. In some sense, it is a redoing, yes. But such a redoing that solves a specific problem and also gives us some potential benefits, listed above. We can discuss it tomorrow, if you'd like, I can elaborate on that. 

Anyway, thank you for spotting this, it was helpful.

**Pull request merged**